---
layout: post
title: Library 3.17 Algorithmic Efficiency Homework
author: Luke-Sanders
permalink: /homework/3-17/
---

## Popcorn Hack

In [7]:
import math

n = 8
linear_checks = n
binary_checks = 0 if n == 0 else math.floor(math.log2(n)) + 1
print(n)
print(linear_checks)
print(binary_checks)

8
8
4


In [8]:
import math

n = 16
linear_checks = n
binary_checks = 0 if n == 0 else math.floor(math.log2(n)) + 1
print(n)
print(linear_checks)
print(binary_checks)

16
16
5


Prediction: For 8 books, linear search needs 8 checks in the worst case and binary search needs 4.

Actual output (n = 8): 8, 8, 4.

Actual output (n = 16): 16, 16, 5.

Explanation: When I doubled n from 8 to 16, the linear count doubled (8 to 16), but the binary count only went up by one (4 to 5). Binary search cuts the catalog in half each time, so doubling the catalog only adds one more check.

Recommendation: For a big catalog that is already sorted, use binary search because it needs far fewer checks.

## MCQ

MCQ 3.17: 2/4 | answers: A, B, B, C

1. Binary search requires: **A (a sorted catalog with direct access to entries)**.
2. Which helps compare two correct searches? **B (work and memory as input grows)**.
3. A quadratic model uses 100 checks at n = 10. At n = 20 it uses: **B (400)**, because doubling n gives four times the work.
4. A short-books-first reading heuristic: **C (can be useful without guaranteeing the best list)**.

## Homework

Each catalog has n books. Scanning every book takes n visits. Comparing every book with every book takes n * n ordered pairs.

In [9]:
n = 12
visits = n
pairs = n * n
print(n)
print(visits)
print(pairs)

12
12
144


Prediction: 12, 12, 144.

Actual output: 12, 12, 144.

In [10]:
n = 24
visits = n
pairs = n * n
print(n)
print(visits)
print(pairs)

24
24
576


Prediction: 24, 24, 576.

Actual output: 24, 24, 576.

Explanation of growth: When I doubled the catalog from 12 to 24 books, the visits doubled (12 to 24) but the pairs quadrupled (144 to 576). Visits grow in a straight line with n. Pairs grow much faster because every book is compared with every book, which is n times n.

In [11]:
import math

n = 1000
linear_checks = n
binary_checks = 0 if n == 0 else math.floor(math.log2(n)) + 1
print(n)
print(linear_checks)
print(binary_checks)

1000
1000
10


Actual output: 1000, 1000, 10.

Recommendation: For repeated searches of a catalog that is already sorted, the librarian should use binary search. It needs at most 10 checks instead of 1,000.

What changes if it is not sorted: Binary search only works on sorted data. If the catalog is unsorted, it has to be sorted first, and sorting takes extra work. For a single search, linear search might be cheaper.

What changes if a sorted copy must be stored: Keeping a second sorted copy uses extra memory, so there is a tradeoff between time and memory.

## Tests

### Test 1: Main catalog (n = 12)

In [12]:
n = 12
visits = n
pairs = n * n
print(n)
print(visits)
print(pairs)

12
12
144


Prediction: 12, 12, 144.

Actual output: 12, 12, 144.

Explanation: Visits equal n, and pairs equal n times n, so 12 * 12 = 144.

### Test 2: Empty catalog (n = 0)

In [13]:
n = 0
visits = n
pairs = n * n
print(n)
print(visits)
print(pairs)

0
0
0


Prediction: 0, 0, 0.

Actual output: 0, 0, 0.

Explanation: With no books there is nothing to visit and nothing to pair, so all the counts are zero. Growth patterns like doubling do not matter here because there is no work at all.

## Design Thinking

Librarian need: A librarian wants to find a book quickly, even when the catalog has thousands of entries.

My goal: Compare ways to search by counting the checks each one needs, instead of judging by how short the code is.

Methods considered: I compared linear search (check one at a time) and binary search (repeatedly halve a sorted catalog).

Prototype: A small calculator that shows the book count, the visits, and the ordered pairs.

Test and revision: I tested n = 12, n = 24, and n = 0. The empty catalog test showed me that all the counts should be zero, so I checked that my code did not break with n = 0.

Clear labels: The output would be easier to read with labels like "Books", "Visits", and "Pairs", so the librarian does not need to understand the formulas.